### ISSUE TITLE
Monitor names are swapped on X11

### ISSUE THREAD

**Issue by WieeRd** [nwg-piotr/azote#202]
**Describe the bug**

I have 2 monitors connected, HDMI-0 and DP-0.

```
❯ xrandr | rg " connected "
HDMI-0 connected 1080x1920+2560+0 right (normal left inverted right x axis y axis) 527mm x 296mm
DP-0 connected primary 2560x1440+0+0 (normal left inverted right x axis y axis) 597mm x 336mm
```

![image](https://github.com/user-attachments/assets/625e26cc-289c-46d0-b90a-b1e082b309fa)

**The above screenshot was captured on DP-0, the 2560x1440 one.**
For some reason, the monitor names seem to be swapped with each other.
When I set a wallpaper for HDMI-0 it is applied to DP-0 and vice versa.

```
❯ xrandr --output HDMI-0 --off
❯ xrandr | rg " connected "
HDMI-0 connected (normal left inverted right x axis y axis)
DP-0 connected primary 2560x1440+0+0 (normal left inverted right x axis y axis) 597mm x 336mm
```

![image](https://github.com/user-attachments/assets/7ebfb48e-b1c4-4f4c-936b-ce2a0e5d9aa5)

If I disable HDMI-0 and have only DP-0 enabled, only HDMI-0 is shown in Azote.
And of course, it is actually linked with DP-0.

I have yet to test what happens if I plug in 3 monitors because I don't have a third one.
Perhaps is there an off-by-one indexing error for getting the monitor names?

**Desktop (please complete the following information):**
 - Linux distribution: Arch
 - WM: Qtile (X11)

**Azote version (please state it clearly if you use -git version):**
 - 1.13.1

**Comment by WieeRd** [comment]
TL;DR - I have DP-0 (2560x1440) and HDMI-0 (1920x1080).
When both are enabled, both the name and resolution are swapped and shown as HDMI-0 (1920x1080) and DP-0 (2560x1440).
When only DP-0 is enabled, only the display name is wrong and shown as HDMI-0 (2560x1440).

This is very strange...

**Comment by nwg-piotr** [comment]
LOL, you made me open Openbox first time in a couple of years. Well, if configured well with arandr, outputs show up properly on my side. I have no idea what goes wrong on your machine. I abandoned X11 years ago.

![2024-11-22-004517_5560x1920_scrot](https://github.com/user-attachments/assets/99c89112-e44e-42bf-b4f5-182f23d9074d)

**Comment by WieeRd** [comment]
As a X11 user I can confirm I don't remember what happiness feels like.
This bug doesn't really hinder the features of Azote too much since the wallpaper settings still work.
However it is such a weird issue that I'm willing to track it down just to see what caused it, out of curiosity.

Any idea where to take a look or ways to debug this? None of the other tools such as arandr or wallpaper / monitor configuration tools from KDE/GNOME seem to report monitor names incorrectly.

**Comment by nwg-piotr** [comment]
On X11 Azote detects monitors on the basis of the `xrandr` output, see: https://github.com/nwg-piotr/azote/blob/9ced6a35c6e315af113830522c25b9af7bba3411/azote/tools.py#L235

The code is old, untouched for years, and could be optimized, but it should probably work well.

**Comment by WieeRd** [comment]
It shouldn't be too hard to fix, I'll return with a PR during the weekends. The surrounding code could also use some work.

**Comment by WieeRd** [comment]
@nwg-piotr Do you mind adding an extra optional dependency, [`python-xlib`](https://github.com/python-xlib/python-xlib)?

Dealing with the `xrandr` output was painful, dirty, and still unsure if I've correctly implemented it. It probably isn't. Probably the reason why you wrote once and never wanted to read for years.

On the other hand, this is how clean it gets if we use a dedicated library instead.

```python
from Xlib import display

display = display.Display()
root = display.screen().root
for i, m in enumerate(root.xrandr_get_monitors().monitors):
    print({
        "name": display.get_atom_name(m.name),
        "x": m.x,
        "y": m.y,
        "width": m.width_in_pixels,
        "height": m.height_in_pixels,
        "xrandr-idx": i,
    })
```

`python-xlib` is very much available on [PyPI](https://pypi.org/project/python-xlib/), [Arch](https://archlinux.org/packages/extra/any/python-xlib/), [Ubuntu](https://packages.ubuntu.com/oracular/python3-xlib), [Fedora](https://packages.fedoraproject.org/pkgs/python-xlib/python3-xlib/), and [NixOS Stable](https://search.nixos.org/packages?channel=24.05&from=0&size=50&sort=relevance&type=packages&query=python+xlib).
Many of your nwg-shell tools already indirectly requires `python-xlib` as a dependency of `python-i3ipc` as well.
Also, the `xorg-xrandr` dependency can be dropped. Did I mention how I hate parsing plaintext outputs without well-defined formats?

**Comment by nwg-piotr** [comment]
>> Probably the reason why you wrote once and never wanted to read for years.

Haha, I started this repo right after migrating all my stuff from i3 to sway, back in 2019. I added some X11 support via feh, but never actually used it.

Add the Xlib dependency if you need, but from now on you're responsible for X11-related stuff in this program. ;)
